File: plugins/approvej-intellij-plugin/src/main/java/org/approvej/intellij/ApproveJRefactoringListenerProvider.java
package org.approvej.intellij;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectUtil;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiNamedElement;
import com.intellij.refactoring.listeners.RefactoringElementListener;
import com.intellij.refactoring.listeners.RefactoringElementListenerProvider;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Updates the {@code .approvej/inventory.properties} file after a test class or method is renamed.
 * The listener fires within the rename command, so the inventory update is part of the same
 * undoable operation.
 */
public final class ApproveJRefactoringListenerProvider
    implements RefactoringElementListenerProvider {

  @Override
  public @Nullable RefactoringElementListener getListener(@NotNull PsiElement element) {
    PsiMethod method = ApproveJRenamePsiElementProcessor.resolveMethod(element);
    if (method != null) return methodRenameListener(method);
    PsiClass clazz = ApproveJRenamePsiElementProcessor.resolveClass(element);
    if (clazz != null) return classRenameListener(clazz);
    return null;
  }

  private static @Nullable RefactoringElementListener methodRenameListener(
      @NotNull PsiMethod method) {
    PsiClass containingClass = method.getContainingClass();
    if (containingClass == null
        || containingClass.getQualifiedName() == null
        || containingClass.getName() == null) {
      return null;
    }

    String className = containingClass.getQualifiedName();
    String simpleClassName = containingClass.getName();
    String oldMethodName = method.getName();
    Project project = method.getProject();

    List<VirtualFile> approvedFiles =
        InventoryUtil.findApprovedFiles(className, oldMethodName, project);
    if (approvedFiles.isEmpty()) return null;

    VirtualFile projectDir = ProjectUtil.guessProjectDir(project);
    Map<String, String> oldPaths = captureRelativePaths(approvedFiles, projectDir);
    String oldTestReference = "%s#%s".formatted(className, oldMethodName);

    return new RefactoringElementListener() {
      @Override
      public void elementMoved(@NotNull PsiElement newElement) {}

      @Override
      public void elementRenamed(@NotNull PsiElement newElement) {
        String newMethodName =
            ApproveJRenamePsiElementProcessor.stripBackticks(
                ((PsiNamedElement) newElement).getName());

        Map<String, String> pathRenames = new HashMap<>();
        for (var entry : oldPaths.entrySet()) {
          String oldPath = entry.getKey();
          String oldFileName = entry.getValue();
          String newFileName =
              ApproveJRenamePsiElementProcessor.computeNewFileName(
                  oldFileName, simpleClassName, oldMethodName, newMethodName);
          if (newFileName != null) {
            pathRenames.put(
                oldPath,
                oldPath.substring(0, oldPath.length() - oldFileName.length()) + newFileName);
          }
        }

        String newTestReference = "%s#%s".formatted(className, newMethodName);
        InventoryUtil.updateEntries(
            project, pathRenames, Map.of(oldTestReference, newTestReference));
      }
    };
  }

  private static @Nullable RefactoringElementListener classRenameListener(@NotNull PsiClass clazz) {
    String qualifiedName = clazz.getQualifiedName();
    String oldClassName = clazz.getName();
    if (qualifiedName == null || oldClassName == null) return null;

    Project project = clazz.getProject();
    VirtualFile projectDir = ProjectUtil.guessProjectDir(project);

    Map<String, String> allOldPaths = new HashMap<>();
    Map<String, String> oldTestReferences = new HashMap<>();

    for (PsiMethod method : clazz.getMethods()) {
      String methodName = method.getName();
      List<VirtualFile> approvedFiles =
          InventoryUtil.findApprovedFiles(qualifiedName, methodName, project);
      if (approvedFiles.isEmpty()) continue;

      allOldPaths.putAll(captureRelativePaths(approvedFiles, projectDir));
      oldTestReferences.put("%s#%s".formatted(qualifiedName, methodName), methodName);
    }

    if (allOldPaths.isEmpty()) return null;

    return new RefactoringElementListener() {
      @Override
      public void elementMoved(@NotNull PsiElement newElement) {}

      @Override
      public void elementRenamed(@NotNull PsiElement newElement) {
        String newClassName = ((PsiNamedElement) newElement).getName();
        String packageName =
            qualifiedName.substring(0, qualifiedName.length() - oldClassName.length());
        String newQualifiedName = packageName + newClassName;

        Map<String, String> pathRenames = new HashMap<>();
        for (var entry : allOldPaths.entrySet()) {
          String oldPath = entry.getKey();
          String oldFileName = entry.getValue();

          if (oldFileName.startsWith(oldClassName + "-")) {
            String newFileName = newClassName + oldFileName.substring(oldClassName.length());
            pathRenames.put(
                oldPath,
                oldPath.substring(0, oldPath.length() - oldFileName.length()) + newFileName);
          } else {
            // Subdirectory pattern: path contains /OldClass/filename
            String oldDirSegment = "/" + oldClassName + "/";
            String newDirSegment = "/" + newClassName + "/";
            int dirIndex = oldPath.indexOf(oldDirSegment);
            if (dirIndex >= 0) {
              pathRenames.put(
                  oldPath, oldPath.substring(0, dirIndex) + newDirSegment + oldFileName);
            }
          }
        }

        Map<String, String> testReferenceRenames = new HashMap<>();
        for (var entry : oldTestReferences.entrySet()) {
          String oldRef = entry.getKey();
          String methodName = entry.getValue();
          testReferenceRenames.put(oldRef, "%s#%s".formatted(newQualifiedName, methodName));
        }

        InventoryUtil.updateEntries(project, pathRenames, testReferenceRenames);
      }
    };
  }

  private static @NotNull Map<String, String> captureRelativePaths(
      @NotNull List<VirtualFile> files, @Nullable VirtualFile projectDir) {
    Map<String, String> paths = new HashMap<>();
    if (projectDir == null) return paths;
    for (VirtualFile file : files) {
      String relativePath = VfsUtil.getRelativePath(file, projectDir);
      if (relativePath != null) {
        paths.put(relativePath, file.getName());
      }
    }
    return paths;
  }
}

File: plugins/approvej-intellij-plugin/src/main/kotlin/org/approvej/intellij/ApproveJRefactoringListenerProvider.kt
package org.approvej.intellij

import com.intellij.openapi.project.guessProjectDir
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiMethod
import com.intellij.psi.PsiNamedElement
import com.intellij.refactoring.listeners.RefactoringElementListener
import com.intellij.refactoring.listeners.RefactoringElementListenerProvider

/**
 * Updates the `.approvej/inventory.properties` file after a test class or method is renamed. The
 * listener fires within the rename command, so the inventory update is part of the same undoable
 * operation.
 */
class ApproveJRefactoringListenerProvider : RefactoringElementListenerProvider {

  override fun getListener(element: PsiElement): RefactoringElementListener? {
    val method = ApproveJRenamePsiElementProcessor.resolveMethod(element)
    if (method != null) return methodRenameListener(method)
    val clazz = ApproveJRenamePsiElementProcessor.resolveClass(element)
    if (clazz != null) return classRenameListener(clazz)
    return null
  }

  companion object {
    private fun methodRenameListener(method: PsiMethod): RefactoringElementListener? {
      val containingClass = method.containingClass ?: return null
      val className = containingClass.qualifiedName ?: return null
      val simpleClassName = containingClass.name ?: return null
      val oldMethodName = method.name
      val project = method.project

      val approvedFiles = InventoryUtil.findApprovedFiles(className, oldMethodName, project)
      if (approvedFiles.isEmpty()) return null

      val projectDir = project.guessProjectDir()
      val oldPaths = captureRelativePaths(approvedFiles, projectDir)
      val oldTestReference = "$className#$oldMethodName"

      return object : RefactoringElementListener {
        override fun elementMoved(newElement: PsiElement) {
          // not applicable for method renames
        }

        override fun elementRenamed(newElement: PsiElement) {
          val newMethodName =
            ApproveJRenamePsiElementProcessor.stripBackticks(
              (newElement as PsiNamedElement).name ?: return
            )

          val pathRenames =
            computeMethodPathRenames(oldPaths, simpleClassName, oldMethodName, newMethodName)
          val newTestReference = "$className#$newMethodName"
          InventoryUtil.updateEntries(
            project,
            pathRenames,
            mapOf(oldTestReference to newTestReference),
          )
        }
      }
    }

    private fun classRenameListener(clazz: PsiClass): RefactoringElementListener? {
      val qualifiedName = clazz.qualifiedName ?: return null
      val oldClassName = clazz.name ?: return null
      val project = clazz.project
      val projectDir = project.guessProjectDir()

      val allOldPaths = mutableMapOf<String, String>()
      val oldTestReferences = mutableMapOf<String, String>()

      for (method in clazz.methods) {
        val methodName = method.name
        val approvedFiles = InventoryUtil.findApprovedFiles(qualifiedName, methodName, project)
        if (approvedFiles.isEmpty()) continue

        allOldPaths.putAll(captureRelativePaths(approvedFiles, projectDir))
        oldTestReferences["$qualifiedName#$methodName"] = methodName
      }

      if (allOldPaths.isEmpty()) return null

      return object : RefactoringElementListener {
        override fun elementMoved(newElement: PsiElement) {
          // not applicable for class renames
        }

        override fun elementRenamed(newElement: PsiElement) {
          val newClassName = (newElement as PsiNamedElement).name ?: return
          val packageName = qualifiedName.substring(0, qualifiedName.length - oldClassName.length)
          val newQualifiedName = packageName + newClassName

          val pathRenames = computeClassPathRenames(allOldPaths, oldClassName, newClassName)
          val testReferenceRenames =
            oldTestReferences.entries.associate { (oldRef, methodName) ->
              oldRef to "$newQualifiedName#$methodName"
            }

          InventoryUtil.updateEntries(project, pathRenames, testReferenceRenames)
        }
      }
    }

    private fun computeMethodPathRenames(
      oldPaths: Map<String, String>,
      simpleClassName: String,
      oldMethodName: String,
      newMethodName: String,
    ): Map<String, String> {
      return buildMap {
        for ((oldPath, oldFileName) in oldPaths) {
          val newFileName =
            ApproveJRenamePsiElementProcessor.computeNewFileName(
              oldFileName,
              simpleClassName,
              oldMethodName,
              newMethodName,
            )
          if (newFileName != null) {
            put(oldPath, oldPath.substring(0, oldPath.length - oldFileName.length) + newFileName)
          }
        }
      }
    }

    private fun computeClassPathRenames(
      allOldPaths: Map<String, String>,
      oldClassName: String,
      newClassName: String,
    ): Map<String, String> {
      return buildMap {
        for ((oldPath, oldFileName) in allOldPaths) {
          if (oldFileName.startsWith("$oldClassName-")) {
            val newFileName = newClassName + oldFileName.substring(oldClassName.length)
            put(oldPath, oldPath.substring(0, oldPath.length - oldFileName.length) + newFileName)
          } else {
            val oldDirSegment = "/$oldClassName/"
            val dirIndex = oldPath.indexOf(oldDirSegment)
            if (dirIndex >= 0) {
              put(oldPath, oldPath.substring(0, dirIndex) + "/$newClassName/" + oldFileName)
            }
          }
        }
      }
    }

    private fun captureRelativePaths(
      files: List<VirtualFile>,
      projectDir: VirtualFile?,
    ): Map<String, String> {
      if (projectDir == null) return emptyMap()
      return buildMap {
        for (file in files) {
          val relativePath = VfsUtil.getRelativePath(file, projectDir)
          if (relativePath != null) {
            put(relativePath, file.name)
          }
        }
      }
    }
  }
}
